File: 2/js/p1.js
const readline = require('readline');

let position = {
  horizontal: 0,
  vertical: 0
}

const movements = {
  forward: {
    mod: 1,
    pos: 'horizontal'
  },
  down: {
    mod: 1,
    pos: 'vertical'
  },
  up: {
    mod: -1,
    pos: 'vertical'
  }
}

const regexp = /(forward|up|down)\s+(\d+)/;

const rl = readline.createInterface({
  input: process.stdin
});

rl.on('line', (l) => {
  const [, direction, power] = l.match(regexp);

  const move = movements[direction];

  position[move.pos] += move.mod * power;
});

rl.on('close', () => {
  console.log(`(${position.horizontal}, ${position.vertical}) ${position.horizontal * position.vertical}`);
});

File: 2/lua/p1.lua
local position = {
  horizontal = 0,
  depth = 0
}

for line in io.lines() do
  local direction, power = line:match('(%a+)%s+(%d+)')
  power = tonumber(power)

  if direction == 'forward' then
    position.horizontal = position.horizontal + power
  elseif direction == 'up' then
    position.depth = position.depth - power
  elseif direction == 'down' then
    position.depth = position.depth + power
  end
end

print(position.horizontal, position.depth, position.horizontal * position.depth)
